File: ios/Dash/Core/Contracts/AgentContracts.swift
import Foundation

enum RegisteredAgentStatus: String, Codable, Hashable, Sendable {
  case registered
  case active
  case disabled
}

struct AgentSkillsDTO: Codable, Hashable, Sendable {
  let paths: [String]?
  let urls: [String]?
}

struct AgentSwarmDTO: Codable, Hashable, Sendable {
  let enabled: Bool?
  let maxConcurrentWorkers: Int?
  let maxWorkersPerRun: Int?
  let maxSteersPerWorker: Int?
  let maxRunSeconds: Int?
  let allowedModels: [String]?
}

struct AgentConfigDTO: Codable, Hashable, Sendable {
  let name: String
  let model: String
  let systemPrompt: String
  let fallbackModels: [String]?
  let tools: [String]?
  let skills: AgentSkillsDTO?
  let workspace: String?
  let maxTokens: Int?
  let mcpServers: [String]?
  let swarm: AgentSwarmDTO?
  let plugins: [String]?
  let providers: [String]?
}

/// A skill as the mobile API exposes it: read-only, without the gateway's
/// on-disk path. `source` says where it came from — `agent` means the agent
/// wrote it for itself.
struct SkillDTO: Codable, Hashable, Identifiable, Sendable {
  var id: String { name }
  let name: String
  let description: String
  let trigger: String?
  let source: SkillSource
  let content: String?
}

/// Unknown sources decode to `.unknown` rather than throwing, so a gateway
/// that adds one does not break skill listing on an older app.
enum SkillSource: String, Codable, Hashable, Sendable {
  case managed
  case agent
  case remote
  case plugin
  case unknown

  init(from decoder: Decoder) throws {
    let raw = try decoder.singleValueContainer().decode(String.self)
    self = SkillSource(rawValue: raw) ?? .unknown
  }

  /// How the source reads to a user.
  var label: String {
    switch self {
    case .agent: return "Learned"
    case .managed: return "Added"
    case .remote: return "Installed"
    case .plugin: return "Built-in"
    case .unknown: return "Other"
    }
  }
}

struct RegisteredAgentDTO: Codable, Hashable, Identifiable, Sendable {
  let id: String
  let name: String
  let config: AgentConfigDTO
  let status: RegisteredAgentStatus
  let registeredAt: Date
}

struct CreateAgentRequest: Codable, Hashable, Sendable {
  let name: String
  let model: String
  let systemPrompt: String
}

struct UpdateAgentRequest: Codable, Hashable, Sendable {
  let model: String?
  let systemPrompt: String?

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encodeIfPresent(model, forKey: .model)
    try container.encodeIfPresent(systemPrompt, forKey: .systemPrompt)
  }
}

enum ModelsSource: String, Codable, Hashable, Sendable {
  case live
  case bootstrap
}

struct ModelDTO: Codable, Hashable, Sendable {
  let value: String
  let label: String
  let provider: String
}

struct ModelsResponseDTO: Codable, Hashable, Sendable {
  let models: [ModelDTO]
  let source: ModelsSource
  let errors: [String: String]
  let fetchedAt: Date
  let supportedModelsReviewedAt: String
}

/// A memory row as the mobile memory routes serve it. `type` reuses
/// `MemoryTypeDTO` (declared in `AgentEvent.swift`, where the `memory_saved`
/// event needs it) so the buckets can never drift apart. `createdAt` and
/// `updatedAt` are bare `YYYY-MM-DD` days on the wire, not RFC 3339
/// timestamps, so they stay `String` — `ContractCoding`'s date strategy
/// would reject them.
struct MemoryInfoDTO: Codable, Hashable, Identifiable, Sendable {
  var id: String { name }
  let name: String
  let description: String
  let type: MemoryTypeDTO
  let source: String
  let createdAt: String
  let updatedAt: String
  let size: Int
}

struct MemoryDeleteResponseDTO: Codable, Hashable, Sendable {
  let name: String
}

File: android/core/contracts/src/main/kotlin/app/dash/core/contracts/AgentContracts.kt
@file:UseSerializers(JsonIntegralLongSerializer::class)

package app.dash.core.contracts

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

@Serializable
data class MobileAgent(
  val id: String,
  val name: String,
  val config: MobileAgentConfig,
  val status: AgentStatus,
  val registeredAt: String,
)

@Serializable
data class MobileAgentConfig(
  val name: String,
  val model: String,
  val systemPrompt: String,
  val fallbackModels: List<String>? = null,
  val tools: List<String>? = null,
  val skills: AgentSkills? = null,
  val workspace: String? = null,
  val maxTokens: Long? = null,
  val mcpServers: List<String>? = null,
  val plugins: List<String>? = null,
  val providers: List<String>? = null,
  val swarm: AgentSwarm? = null,
)

@Serializable
data class AgentSkills(val paths: List<String>? = null, val urls: List<String>? = null)

@Serializable
data class AgentSwarm(
  val enabled: Boolean? = null,
  val maxConcurrentWorkers: Long? = null,
  val maxWorkersPerRun: Long? = null,
  val maxSteersPerWorker: Long? = null,
  val maxRunSeconds: Long? = null,
  val allowedModels: List<String>? = null,
)

@Serializable
enum class AgentStatus {
  @SerialName("registered")
  Registered,

  @SerialName("active")
  Active,

  @SerialName("disabled")
  Disabled,
}

@Serializable
data class CreateMobileAgentRequest(
  val name: String,
  val model: String,
  val systemPrompt: String,
)

@Serializable
data class UpdateMobileAgentRequest(
  val model: String? = null,
  val systemPrompt: String? = null,
)

@Serializable
data class MobileActionResponse(val ok: Boolean) {
  init {
    require(ok) { "MobileActionResponse.ok must be true" }
  }
}

@Serializable(with = MobileSkillSourceSerializer::class)
sealed interface MobileSkillSource {
  data object Managed : MobileSkillSource

  data object Agent : MobileSkillSource

  data object Remote : MobileSkillSource

  data object Plugin : MobileSkillSource

  data class Unknown(val wireValue: String) : MobileSkillSource
}

object MobileSkillSourceSerializer : KSerializer<MobileSkillSource> {
  override val descriptor = PrimitiveSerialDescriptor("MobileSkillSource", PrimitiveKind.STRING)

  override fun deserialize(decoder: Decoder): MobileSkillSource = when (val value = decoder.decodeString()) {
    "managed" -> MobileSkillSource.Managed
    "agent" -> MobileSkillSource.Agent
    "remote" -> MobileSkillSource.Remote
    "plugin" -> MobileSkillSource.Plugin
    else -> {
      if (value.isEmpty()) throw SerializationException("MobileSkill.source must be nonempty")
      MobileSkillSource.Unknown(value)
    }
  }

  override fun serialize(encoder: Encoder, value: MobileSkillSource) {
    val wireValue = when (value) {
      MobileSkillSource.Managed -> "managed"
      MobileSkillSource.Agent -> "agent"
      MobileSkillSource.Remote -> "remote"
      MobileSkillSource.Plugin -> "plugin"
      is MobileSkillSource.Unknown -> value.wireValue
    }
    if (wireValue.isEmpty()) throw SerializationException("MobileSkill.source must be nonempty")
    encoder.encodeString(wireValue)
  }
}

@Serializable
data class MobileSkill(
  val name: String,
  val description: String,
  val trigger: String? = null,
  val source: MobileSkillSource,
  val content: String? = null,
)

@Serializable
data class MobileMemoryInfo(
  val name: String,
  val description: String,
  val type: String,
  val source: String,
  val createdAt: String,
  val updatedAt: String,
  val size: Long,
)

@Serializable
data class MobileMemoryRecord(
  val name: String,
  val description: String,
  val type: String,
  val source: String,
  val createdAt: String,
  val updatedAt: String,
  val content: String,
)

@Serializable
data class MobileMemoryDeleteResponse(val name: String)

@Serializable
data class MobileModel(val value: String, val label: String, val provider: String)

@Serializable
data class MobileModelsResponse(
  val models: List<MobileModel>,
  val source: String,
  val errors: Map<String, String>,
  val fetchedAt: String,
  val supportedModelsReviewedAt: String,
)

@Serializable(with = AgentEventSerializer::class)
data class AgentEvent(val type: String, val raw: JsonObject)

object AgentEventSerializer : KSerializer<AgentEvent> {
  override val descriptor = buildClassSerialDescriptor("AgentEvent")

  override fun deserialize(decoder: Decoder): AgentEvent {
    val raw = decoder.decodeSerializableValue(JsonObject.serializer())
    val type = raw["type"] as? JsonPrimitive
    if (type == null || !type.isString || type.content.isEmpty()) {
      throw SerializationException("AgentEvent.type must be a nonempty string")
    }
    WireRules.validateAgentEvent(raw)
    return AgentEvent(type.content, raw)
  }

  override fun serialize(encoder: Encoder, value: AgentEvent) {
    val rawType = value.raw["type"] as? JsonPrimitive
    if (
      value.type.isEmpty() ||
      rawType == null ||
      !rawType.isString ||
      rawType.content != value.type
    ) {
      throw SerializationException("AgentEvent.type must match its raw object")
    }
    WireRules.validateAgentEvent(value.raw)
    encoder.encodeSerializableValue(JsonObject.serializer(), value.raw)
  }
}
